src/ExtraValuesAssigner.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+namespace Wikibase\DataModel;
+
+/**
+ * Assigns extra values to the serialization of an object
+ * based on the given extra values providers.
+ *
+ * @since 2.1
+ *
+ * @licence GNU GPL v2+
+ * @author Bene* < benestar.wikimedia@gmail.com >
+ */
+class ExtraValuesAssigner {
+
+	/**
+	 * @var ExtraValuesProvider[]
+	 */
+	private $extraValuesProviders;
+
+	/**
+	 * @param ExtraValuesProvider[] $extraValuesProviders
+	 */
+	public function __construct( array $extraValuesProviders = array() ) {
+		$this->extraValuesProviders = $extraValuesProviders;
+	}
+
+	/**
+	 * @param array $serialization
+	 * @param mixed $object
+	 * @return array
+	 */
+	public function addExtraValues( $serialization, $object ) {
+		foreach ( $this->extraValuesProviders as $extraValuesProvider ) {
+			if ( $extraValuesProvider->isProviderFor( $object ) ) {
+				$serialization = array_merge(
+					$extraValuesProvider->getExtraValues( $object ),
+					$serialization
```

Comment:
But that would still require hardcoding the paths of the serialization in some other place outside of this component, right? While I think your idea is definitely worth consideration, it would imo be nicer to not hardcode paths in the serialization. Also this would even more create a "parallel serializer hierarchy outside of the original serializer hierarchy".

In theory when using this serialization component we should only know that the result is an associative array that can be passed to json_encode but nothing more about the content of this array.